pkg/apis/core/validation/validation.go

```diff
@@ -3667,6 +3680,10 @@ func ValidatePodSecurityContext(securityContext *core.PodSecurityContext, spec *
 			allErrs = append(allErrs, validateSysctls(securityContext.Sysctls, fldPath.Child("sysctls"))...)
 		}
 
+		if securityContext.FSGroupChangePolicy != nil {
```

Comment:
should we check only if feature is enabled, we allow to set the policy?

Reply:
We depend on the logic [here](https://github.com/kubernetes/kubernetes/pull/88488/commits/de97ba6b06f925b3d34d5ad28fe51c5ae9e024e3#diff-3e0068b76443e20a2e4bcc68b91a9f45R500) to drop the field if the feature is disabled. The reason for this is to be able to support rollback or disabling the feature. In that case, the field exists on the object because it was created when the feature was turned on. We don't want to fail updates or deletion to the object if the feature got disabled.

Reply:
So during pod creation, if the FSGroupChangePolicy is disabled, we allow user to set FSGroupChangePolicy, but drop it silently without any warning or error msg? Will this cause some confusion to user if they did not know this field is dropped?
How about the following?
1. During pod creation, if FSGroupChangePolicy is disabled, if user set FSGroupChangePolicy, validation will fail.
2. During pod update,  if FSGroupChangePolicy is disabled, if FSGroupChangePolicy is disabled, drop this field (but it would be better to give some warning information about it?)

Also notice another PR that might has the similar issue? https://github.com/kubernetes/kubernetes/pull/84051#pullrequestreview-366588225

Reply:
I agree it's not a great user experience to just silently drop the field, but dropping the field is the current process for [adding new fields](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#alpha-field-in-existing-api-version)

Reply:
The semantic is "as if the field did not exist".  We drop non-existent fields silently, rather than error, so when the gate is disabled, we must do the same.

Reply:
I am wondering why this behavior is desirable? From customer point of view, they set the the policy. But cluster admin set the feature gate and they might not aware it. It is might not easy for customer to discover this hidden behavior.

Reply:
If they set the policy but cluster admin has not upgraded the cluster, the field will get dropped, silently.

If they set the policy but cluster admin has not enabled the feature, the field will get dropped, silently.

Reply:
Right, but is there any benefit of dropping the fields silently instead of failing on errors during podCreate validation when adding new fields for a feature? 

As @msau42 pointed out, in case of roll back, if the field is set and cluster is downgraded or feature is disabled, it needs to allow dropping this field during podUpdate to avoid failures. 

I am trying to compare between the current process and the one I mentioned https://github.com/kubernetes/kubernetes/pull/88488/files#r385848872. I feel failing on errors seems better than dropping it silently for costumers to catch the problem. But there might be concerns about inconsistency for validation between podCreate and podUpdate?

Reply:
I think concern is - difference in behaviour that clients have to deal with. For example:

1. Client talking to 1.17 api-server sets a invalid field. The field gets silently dropped because api-server does not know about it.
2. Client talking to 1.18 api-server sets a invalid field (but now behind a alpha-feature gate) will suddenly start getting errors even though feature-gate is disabled.

We want to preserve client's expectation between 1.17 and 1.18 field behind an alpha feature-gate.

Reply:
> I think concern is - difference in behaviour that clients have to deal with. For example:
> 
> 1. Client talking to 1.17 api-server sets a invalid field. The field gets silently dropped because api-server does not know about it.
An invalid fields will be dropped silently? I thought validation will fail on it?
> 2. Client talking to 1.18 api-server sets a invalid field (but now behind a alpha-feature gate) will suddenly start getting errors even though feature-gate is disabled.
> 
> We want to preserve client's expectation between 1.17 and 1.18 field behind an alpha feature-gate.